### ISSUE TITLE
investigate the `reportImportCycles` rule

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#827]
possible improvements to investigate:
- can it report the error on an import statement instead of just the first line of the file?
- can we make it more deterministic?
   - currently it can change what file it's reported in depending on what file you run pyright on
   - this would probably be hard to do. perhaps a hacky solution would be to just report the error in all effected files. this would be noisy but i think making it more deterministic is more important

**Comment by lucascolley** [comment]
I bumped into what seemed like a false positive with this rule, but upstream was unhelpful with explaining where the problem lies. https://github.com/microsoft/pyright/issues/9510.

Is the relative import in the module directory mentioned in that issue a problem? If so, what would be the recommended way to avoid doing that? If not, is this a false positive?

**Comment by DetachHead** [comment]
are you able to provide a minimal repro of the error so i can investigate? i tried checking out that repo but i can't seem to get the error to appear. i assume something was changed to work around the error since the branch you linked on the upstream issue seems to no longer exist

**Comment by lucascolley** [comment]
Thanks for taking a look. I got the error to disappear with the following file structure:

<img width="309" alt="image" src="https://github.com/user-attachments/assets/2cc095a7-1d82-4a7e-8c43-6d25f2059071">

Previously, `_utils.py` was in the same directory as `_funcs.py`. The purported import cycle was:

1. `__init__.py` imports from  `_funcs.py` (true)
2. In `_funcs.py`: `from . import _utils`

Does step (2) perform an import from `__init__.py`? It is a relative import, rather than importing anything that is written in the `__init__.py` file, but I don't know how that works under the hood. At least at runtime, there is no error.

**Comment by KotlinIsland** [comment]
> Does step (2) perform an import from `__init__.py`

```py
import x.y.z
```
will import:
- `x.__init__`
- `x.y.__init__`
- `x.y.z` or `x.y.z.__init__` if it is a package

if `.x` is importing from `.`, then `.__init__` would have always been imported already

> At least at runtime, there is no error.

an import cycle is not a direct indication of a runtime error, although it can signal that one could arise. it is useful as a lint style rule for structuring code in a clean way

**Comment by DetachHead** [comment]
found out today that `reportImportCycles` seems to be limited to only ever report a maximum of 4 instances of the diagnostic per-file, which can be quite confusing when moving code around to remove circular dependencies, because when the diagnostics move between files it can make it look like a new circular dependency was added when it was previously just not being reported due to this limit

https://github.com/DetachHead/basedpyright/blob/dd2ebd8bb38c32fe6124617c7628d92e39b78e46/packages/pyright-internal/src/analyzer/sourceFile.ts#L59
